core/src/main/java/com/github/phantazmnetwork/core/gui/GuiItem.java

```diff
@@ -0,0 +1,181 @@
+         */
+        public @NotNull Builder withReplaceHandler(@NotNull ReplaceHandler replaceHandler) {
+            replaceHandlers.add(Objects.requireNonNull(replaceHandler, "replaceHandler"));
+            return this;
+        }
+
+        /**
+         * Specifies the initial {@link ItemStack}. This may be dynamically updated by specifying an {@link ItemUpdater}
+         * in {@link Builder#withUpdater(ItemUpdater)}.
+         *
+         * @param itemStack the initial ItemStack
+         * @return this instance, for chaining
+         */
+        public @NotNull Builder withItem(@NotNull ItemStack itemStack) {
+            this.itemStack = Objects.requireNonNull(itemStack, "itemStack");
+            return this;
+        }
+
+        /**
+         * Specifies an {@link ItemUpdater} which can be used to periodically update the {@link ItemStack} representing
+         * this GuiItem.
+         *
+         * @param updater the ItemUpdater to use
+         * @return this instance, for chaining
+         */
+        public @NotNull Builder withUpdater(@NotNull ItemUpdater updater) {
+            this.updater = Objects.requireNonNull(updater);
+            return this;
+        }
+
+        /**
+         * Builds a new {@link GuiItem} instance. Can be called multiple times to produce distinct {@link GuiItem}s
+         * that use the same handlers.
+         *
+         * @return a new GuiItem instance
+         */
+        public @NotNull GuiItem build() {
+            return new GuiItem() {
+                private ItemStack stack = Builder.this.itemStack;
+
+                @Override
+                public @NotNull ItemStack getStack() {
+                    return stack;
+                }
+
+                @Override
+                public void handleClick(@NotNull Gui owner, @NotNull Player player, int slot,
+                        @NotNull ClickType clickType) {
+                    for (ClickHandler clickHandler : clickHandlers) {
+                        clickHandler.handleClick(owner, player, slot, clickType);
+                    }
+                }
+
+                @Override
+                public void onRemove(@NotNull Gui owner, int slot) {
+                    for (RemoveHandler removeHandler : removeHandlers) {
+                        removeHandler.onRemove(owner, slot);
+                    }
+                }
+
+                @Override
+                public void onReplace(@NotNull Gui owner, @NotNull GuiItem newItem, int slot) {
+                    for (ReplaceHandler replaceHandler : replaceHandlers) {
+                        replaceHandler.onReplace(owner, newItem, slot);
+                    }
+                }
+
+                @Override
+                public void tick(long time) {
+                    if (updater == null) {
+                        return;
+                    }
+
+                    stack = updater.update(time, stack);
+                }
```

Comment:
generally GuiItem expands upon InventoryObject, with the exception being that the updater is called each tick whereas InventoryObjects require a call to setDirty in order to trigger a redraw in order to save unnecessary item stack updates / calculations, generally I think it would be a good thing to have GuiItems work as InventoryObjects

Reply:
could work, although shouldn't InventoryObjects belong (at least conceptually) to a specific player?

Reply:
no

Reply:
okay, will make GuiItem extend InventoryObject then